subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/install/internal/JdkCacheDirectory.java

```diff
@@ -39,19 +42,26 @@
 
     private final FileOperations operations;
     private final File jdkDirectory;
+    private final FileLockManager lockManager;
 
-    public JdkCacheDirectory(GradleUserHomeDirProvider homeDirProvider, FileOperations operations) {
+    public JdkCacheDirectory(GradleUserHomeDirProvider homeDirProvider, FileOperations operations, FileLockManager lockManager) {
         this.operations = operations;
-        // TODO: [bm] any reason to use CacheScopeMapping instead?
         this.jdkDirectory = new File(homeDirProvider.getGradleUserHomeDirectory(), "jdks");
+        this.lockManager = lockManager;
+        jdkDirectory.mkdir();
     }
 
     public Set<File> listJavaHomes() {
-        final File[] files = jdkDirectory.listFiles();
-        if (files != null) {
-            return Arrays.stream(files).filter(File::isDirectory).map(this::findJavaHome).collect(Collectors.toSet());
+        final FileLock lock = acquireReadLock();
```

Comment:
I am not sure I get the locking you are putting in place.

Looks like two builds cannot download different JDK versions in parallel. And that for a build to know which JDKs are available, a read lock is needed, causing potential contention if a download is happening in parallel.
Is that correct?

Reply:
Downloading in parallel is not supported using the locking strategy I choose. If we want to enable parallel downloads, we'd need to only lock for a specific JDK (which is possible) but it would mean we lose the read lock across all jdks and thus, we'd miss on this scenario:
Build A needs JDK 14, doesn't find it, downloads it (...)
Build B needs JDK 14, doesn't find it, tries to download it. Now we're locking on a specific jdk.

Now we have 2 builds trying downloading the exact same jdk, build B is starved (ok) but actually needs to properly detect that once it has the lock, A downloaded the thing and skip downloading but return the right installation. This was not needed with the read lock across all installations.

I think overall, we can enable a lock strategy per JDK but it might come with some complexity to handle the cases of the same jdk being requested. But we'd win downloading multiple jdks in parallel. WDYT?

Reply:
As discussed, reworked the locking behavior in 5806cc2 to allow parallel downloads and avoid the read lock